src/ucs/sys/sys.c

```diff
@@ -972,23 +976,53 @@ unsigned long ucs_sys_get_pfn(uintptr_t address)
     }
 
     if (pagemap_fd < 0) {
-        return 0; /* could not open file */
+        return UCS_ERR_IO_ERROR; /* could not open file */
     }
 
     offset = (address / ucs_get_page_size()) * sizeof(data);
-    data   = 0;
-    ret    = pread(pagemap_fd, &data, sizeof(data), offset);
+    len    = sizeof(*data) * page_count;
+    data   = ucs_malloc(len, "pfn array");
```

Comment:
added check for NULL
alloca - may fail on large number of requested pages

Reply:
can we use alloca and read few pages every time?
or maybe mmap the file (not sure it's possible)?
let's try to avoid using malloc/free for rcache flows